more information about mqtt performance impact considerations & update a file name (#2170)
theforeman/foreman-documentation: guides/common/modules/proc_rex-pull-based-performance-impact-considerations.adoc

Reviewer: @Imaanpreet I think you can delete this file too, because we have renamed it to "proc_decreasing..."

[id="REX_pull_based_performance_impact_considerations_{context}"]
= REX pull-based transport performance impact considerations

> When {ProjectServer} is configured to use REX pull-based transport for remote execution jobs using "Script" provider, {SmartProxyServer} sends a notification about new job to the client machines via MQTT. This notification does not include actual workload (actual script) the client is supposed to execute. Once a client receives a message about a new remote execution job, it queries the {SmartProxyServer} for actual payload. As the client executes the job, it periodically sends outputs of the job to the capsule, further increasing the number of requests to the {SmartProxyServer}.

Reviewer 2: Please stick to our convention "one sentence per line".

Author: my bad, I forgot. Updated.

> When {ProjectServer} is configured to use REX pull-based transport (see --foreman-proxy-plugin-remote-execution-script-mode installer option) for remote execution jobs using "Commands" job category, {SmartProxyServer} sends a notification about new job to the client machines via MQTT. This notification does not include actual workload (actual script) the client is supposed to execute. Once a client receives a message about a new remote execution job, it queries the {SmartProxyServer} for actual payload. As the client executes the job, it periodically sends outputs of the job to the capsule, further increasing the number of requests to the {SmartProxyServer}.

Reviewer 3: > for remote execution jobs using "Commands" job category

The category is not a good indicator of anything. Maybe something like `for remote execution jobs using "Script" provider` would be better?

Author: Sure, thanks .

> When {ProjectServer} is configured to use REX pull-based transport (see --foreman-proxy-plugin-remote-execution-script-mode installer option) for remote execution jobs using "Commands" job category, {SmartProxyServer} sends a notification about new job to the client machines via MQTT. This notification does not include actual workload (actual script) the client is supposed to execute. Once a client receives a message about a new remote execution job, it queries the {SmartProxyServer} for actual payload. As the client executes the job, it periodically sends outputs of the job to the capsule, further increasing the number of requests to the {SmartProxyServer}.

Reviewer 3: Looks good now

> These requests to the {SmartProxyServer} together with high concurrency allowed by MQTT protocol can cause exhaustion of {SmartProxyServer} available connections and some requests might fail, making some child tasks of remote execution jobs to hang. This also depends on actual job payload, as some jobs are causing additional load to {ProjectServer}, making it compete for resources if you are using integrated {SmartProxyServer}.

Reviewer 2: It's generally "to {SmartProxyServer}" instead of "to _the_ {SmartProxyServer}".

Author: Thanks!

To avoid this, configure your {ProjectServer} and {SmartProxyServer} deployment appropriately with the following parameters:

[options="nowrap", subs="+quotes,verbatim,attributes"]
----
mqtt_ttl - Time interval in seconds given to the host to pick up the job before considering the job undelivered. 
mqtt_rate_limit - Number of jobs that are allowed to run at the same time
mqtt_resend_interval - Time interval in seconds at which the notification should be re-sent to the host until the job is picked up or canceled
----

In order to tune these parameters you can follow the following procedure on the {ProjectServer}:

[options="nowrap", subs="+quotes,verbatim,attributes"]
----

> {foreman-installer} --foreman-proxy-plugin-remote-execution-script-mqtt-rate-limit (value) --foreman-proxy-plugin-remote-execution-script-mqtt-resend-interval (value) --foreman-proxy-plugin-remote-execution-script-mqtt-ttl (value)

Reviewer 2: ```suggestion
{foreman-installer} \
--foreman-proxy-plugin-remote-execution-script-mqtt-rate-limit _My_MQTT_Rate_Limit_ \
--foreman-proxy-plugin-remote-execution-script-mqtt-resend-interval _My_MQTT_Resend_Interval_ \
--foreman-proxy-plugin-remote-execution-script-mqtt-ttl _My_MQTT_Time_To_Live_
```

----

> {SmartProxyServer} logs are in /var/log/foreman-proxy/proxy.log. {SmartProxyServer} uses Webrick HTTP server (no httpd or Puma involved), so there is no simple way to increase its capacity

Reviewer 2: ```suggestion
{SmartProxyServer} logs are in `/var/log/foreman-proxy/proxy.log`.
{SmartProxyServer} uses Webrick HTTP server (no httpd or Puma involved), so there is no simple way to increase its capacity.
```

If you want to mention paths, please use backticks.

I'd also drop things that are not used such as Puma.

> {SmartProxyServer} logs are in /var/log/foreman-proxy/proxy.log. {SmartProxyServer} uses Webrick HTTP server (no httpd or Puma involved), so there is no simple way to increase its capacity

Author: I have just added it for more info @adamruzicka

> {SmartProxyServer} logs are in /var/log/foreman-proxy/proxy.log. {SmartProxyServer} uses Webrick HTTP server (no httpd or Puma involved), so there is no simple way to increase its capacity

Reviewer 3: Fine by me

[id="REX_pull_based_performance_impact_considerations_{context}"]

> = REX pull-based transport performance impact considerations

Author: done. thanks

> When {ProjectServer} is configured to use REX pull-based transport for remote execution jobs using "Script" provider, {SmartProxyServer} sends a notification about new job to the client machines via MQTT.

Reviewer 2: ```suggestion
When {ProjectServer} is configured to use REX pull-based transport for remote execution jobs using "Script" provider, {SmartProxyServer} sends a notification about new job to the client machines through MQTT.
```

> This notification does not include actual workload (actual script) the client is supposed to execute.

Reviewer 2: Is this content partially identical to guides/common/modules/proc_decreasing-performance-impact-of-the-pull-based-rex-transport.adoc?

If so, please either use a snippet or only show it once.

> Once a client receives a message about a new remote execution job, it queries the {SmartProxyServer} for actual payload.

Reviewer 4: ```suggestion
Once a client receives a message about a new remote execution job, it queries the {SmartProxyServer} for the actual payload.
```

> As the client executes the job, it periodically sends outputs of the job to the capsule,further increasing the number of requests to {SmartProxyServer}.

Reviewer 4: ```suggestion
As the client executes the job, it periodically sends outputs of the job to the capsule, further increasing the number of requests to {SmartProxyServer}.
```

These requests to {SmartProxyServer} together with high concurrency allowed by MQTT protocol can cause exhaustion of {SmartProxyServer} available connections and some requests might fail, making some child tasks of remote execution jobs to hang. 
This also depends on actual job payload, as some jobs are causing additional load to {ProjectServer}, making it compete for resources if you are using integrated {SmartProxyServer}.

To avoid this, configure your {ProjectServer} and {SmartProxyServer} deployment appropriately with the following parameters:

> [options="nowrap", subs="+quotes,verbatim,attributes"]
> ----
> mqtt_ttl - Time interval in seconds given to the host to pick up the job before considering the job undelivered. 
> mqtt_rate_limit - Number of jobs that are allowed to run at the same time
> mqtt_resend_interval - Time interval in seconds at which the notification should be re-sent to the host until the job is picked up or canceled
> ----

Reviewer: ```suggestion
* MQTT Time To Live {endash} Time interval in seconds given to the host to pick up the job before considering the job undelivered
* MQTT Rate Limit {endash} Number of jobs that are allowed to run at the same time
* MQTT Resend Interval {endash} Time interval in seconds at which the notification should be re-sent to the host until the job is picked up or cancelled
```

Author: done. thanks

> mqtt_ttl - Time interval in seconds given to the host to pick up the job before considering the job undelivered. 
> mqtt_rate_limit - Number of jobs that are allowed to run at the same time
> mqtt_resend_interval - Time interval in seconds at which the notification should be re-sent to the host until the job is picked up or canceled
> ----

Reviewer 4: ```suggestion
# mqtt_ttl   <1> 
# mqtt_rate_limit   <2>
# mqtt_resend_interval   <3>
----
<1> Time interval in seconds given to the host to pick up the job before considering the job undelivered.
<2> Number of jobs that are allowed to run at the same time.
<3> Time interval in seconds at which the notification should be re-sent to the host until the job is picked up or canceled.
```

> In order to tune these parameters you can follow the following procedure on the {ProjectServer}:

Reviewer 4: ```suggestion
To tune these parameters you can follow the following procedure on the {ProjectServer}:
```

[options="nowrap", subs="+quotes,verbatim,attributes"]
----
{foreman-installer} \
--foreman-proxy-plugin-remote-execution-script-mqtt-rate-limit _My_MQTT_Rate_Limit_ \
--foreman-proxy-plugin-remote-execution-script-mqtt-resend-interval _My_MQTT_Resend_Interval_ \
--foreman-proxy-plugin-remote-execution-script-mqtt-ttl _My_MQTT_Time_To_Live_
----

{SmartProxyServer} logs are in `/var/log/foreman-proxy/proxy.log`. 

> {SmartProxyServer} uses Webrick HTTP server (no httpd or Puma involved), so there is no simple way to increase its capacity

Reviewer 5: If this is a bottleneck, is there an issue opened?

> {SmartProxyServer} uses Webrick HTTP server (no httpd or Puma involved), so there is no simple way to increase its capacity

Reviewer 4: ```suggestion
{SmartProxyServer} uses Webrick HTTP server (no httpd or Puma involved), so there is no simple way to increase its capacity.
```

This also depends on actual job payload, as some jobs are causing additional load to {ProjectServer}, making it compete for resources if you are using integrated {SmartProxyServer}.

To avoid this, configure your {ProjectServer} and {SmartProxyServer} deployment appropriately with the following parameters:

[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# mqtt_ttl   <1> 
# mqtt_rate_limit   <2>
# mqtt_resend_interval   <3>
----
<1> Time interval in seconds given to the host to pick up the job before considering the job undelivered.
<2> Number of jobs that are allowed to run at the same time.
<3> Time interval in seconds at which the notification should be re-sent to the host until the job is picked up or canceled.

To tune these parameters you can follow the following procedure on the {ProjectServer}:

[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# {foreman-installer} \
--foreman-proxy-plugin-remote-execution-script-mqtt-rate-limit _My_MQTT_Rate_Limit_ \
--foreman-proxy-plugin-remote-execution-script-mqtt-resend-interval _My_MQTT_Resend_Interval_ \
--foreman-proxy-plugin-remote-execution-script-mqtt-ttl _My_MQTT_Time_To_Live_
----

> {SmartProxyServer} logs are in `/var/log/foreman-proxy/proxy.log`. 

Reviewer: Perhaps we should tell the user why they should check the logs? Or what to look out for?
